src/herder/HerderImpl.cpp

```diff
@@ -2192,15 +2180,64 @@ HerderImpl::trackingHeartBeat()
 }
 
 void
-HerderImpl::updateTransactionQueue(TxSetFrameConstPtr txSet)
+HerderImpl::updateTransactionQueue(TxSetFrameConstPtr externalizedTxSet)
 {
     ZoneScoped;
+    if (externalizedTxSet == nullptr)
+    {
+        CLOG_DEBUG(Herder,
+                   "No tx set to update tx queue - expected during bootstrap");
+        return;
+    }
+    TxSetFrame::Transactions const* classicTxs = nullptr;
+    TxSetFrame::Transactions const* sorobanTxs = nullptr;
+    TxSetFrame::Transactions txsForNonActiveTxSet;
+
+    // Try to extract the transactions from the tx set.
+    // These transactions are used for tx queue cleanup.
+    if (externalizedTxSet->getResolvedFrame() != nullptr ||
+        externalizedTxSet->previousLedgerHash() ==
+            mApp.getLedgerManager().getLastClosedLedgerHeader().hash)
+    {
+        LedgerTxn ltx(mApp.getLedgerTxnRoot(), false,
+                      TransactionMode::READ_ONLY_WITHOUT_SQL_TXN);
+        auto const* resolvedTxSet = externalizedTxSet->resolve(mApp, ltx);
+        releaseAssert(resolvedTxSet);
+        classicTxs = &resolvedTxSet->getTxsForPhase(TxSetFrame::Phase::CLASSIC);
+        if (resolvedTxSet->numPhases() >
+                static_cast<size_t>(TxSetFrame::Phase::SOROBAN) &&
+            mSorobanTransactionQueue != nullptr)
+        {
+            sorobanTxs =
+                &resolvedTxSet->getTxsForPhase(TxSetFrame::Phase::SOROBAN);
+        }
+    }
+    else
+    {
+        // In some cases the externalized tx set may not be possible to
+        // resolve, e.g. when the node is catching up. In such case we still
+        // may remove the externalized transactions from the queue, but we
+        // need to generate them directly from XDR. We also don't
+        // distinguish phases here for simplicity, as tx queues will just
+        // skip unknown transactions.
+        txsForNonActiveTxSet =
+            externalizedTxSet->createTransactionFrames(mApp.getNetworkID());
+        classicTxs = &txsForNonActiveTxSet;
+        if (mSorobanTransactionQueue != nullptr)
+        {
+            sorobanTxs = &txsForNonActiveTxSet;
```

Comment:
Like I said in my original comment, the point is that tx queue wasn't designed to explicitly handle incorrect types and only expects txs of the right type. Changing this creates new opportunities for bugs. We already have one here in reporting: we count banned transactions twice (both in classic and soroban), which will show up in our dashboards as 100% increase and likely cause confusion

> I don't really want to get into much tx set interpretation outside of resolved tx set frame. I'm actually a bit uncomfortable with createTransactionFrames because it needs to iterate the internal structures and interpreting phases would lead to even more maintenance trouble if we e.g. want to introduce an additional phase.

I don't see the harm in basic interpretation of pure XDR, which is always const and independent of any state. `createTransactionFrames` already goes into the internals of generalized tx set to gather txs, so it's a small change to collect txs in TxSeFrame::TxPhases instead of a single vector

Reply:
I'm a bit confused about the incorrect types - the queue cleanup only really pays attention at the hashes, so that shouldn't cause any trouble. But I guess we don't have much choice here - updated to use phases and removed the resolved case.

Reply:
Switched to XDR-based approach.